python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -330,50 +349,59 @@ def _process_ui():
+
+class AuthState:
+    """
+    Authentication can be in three states. Waiting for authentication to complete,
+    authentication was cancelled (user hit Cancel or CTRL-D) and user provided
+    credentials successfully.
+    """
+    waiting = 1
+    cancelled = 2
+    success = 3
 # When a thread cancels session renewal, this flag is set so other threads know
 # to raise an exception as well.
-_is_authentication_cancelled = False
+_auth_state = AuthState.waiting
 
 
-def _renew_session_internal(user, session_token, credentials_handler):
+def _renew_session_internal(user, credentials_handler):
     """
-    Common login logic, regardless of how we are actually logging in. It will first try to reuse
-    any existing session and if that fails then it will ask for credentials and upon success
-    the credentials will be cached.
-    :raises AuthenticationError: Raised if the authentication is cancelled.
-    :raises AuthenticationDisabled: Raised if authentication was cancelled before.
+    Prompts the user for the password. This method should never be called directly
+    and _renew_session should be called instead.
+    :raises AuthenticationCancelled: Raised if the authentication is cancelled.
     """
     logger.debug("About to take the authentication lock.")
     global _renew_session_internal_lock
     with _renew_session_internal_lock:
-        global _is_authentication_cancelled
-        if _is_authentication_cancelled:
-            raise AuthenticationCancelled()
 
-        # If somebody refreshed the session token on the user object since we tried with
-        # the session token.
-        if user.get_session_token() != session_token:
-            return None
         logger.debug("Took the authentication lock.")
 
+        # When authentication is cancelled, every thread who enter the authentication
+        # critical section should throw as well.
+        global _auth_state
+        if _auth_state == AuthState.cancelled:
+            raise AuthenticationCancelled()
+        # If authentication was successful, simply return.
+        elif _auth_state == AuthState.success:
+            return
+
+        # We're the first thread, so authenticate.
         try:
             logger.debug("Not authenticated, requesting user input.")
-            # Do the actually credentials prompting and authenticating.
             hostname, login, session_token = credentials_handler.authenticate(
                 user.get_host(),
                 user.get_login(),
                 user.get_http_proxy()
             )
+            _auth_state = AuthState.success
+            logger.debug("Login successful!")
+            user.set_session_token(session_token)
         except AuthenticationCancelled:
-            _is_authentication_cancelled = True
+            _auth_state = AuthState.cancelled
             logger.debug("Authentication cancelled")
-            user.clear_saved_user(user.get_host())
+            user.uncache_session_token()
```

Comment:
Hmm, the interface feels like it's moving further away from what I was originally suggesting. It feels like are mixing internals with the public part of the interface, which worries me from a maintenance point of view.

As I have said previously, I think there should be a single User object that representing both shotgun users and scripts. A simple object that can do only a couple things - return details about host, login, proxy, that sort of things - allowing a user to serialize and unpack an object safely, and most importantly request a shotgun instance from it.

By adding `get_session_token`, `set_session_token`, `uncache_session_token`, we are exposing lots of implementation stuff that should really be internal I think. It also creates a bigger distinction between the human user and script user types, where I really think they should just be a single thing!!

Also, why do we need to run `uncache_session_token`? I am not sure that's right. I imagined the workflow like this:

- User is using maya and leaves it open for 24 hours
- User launches App X
- Shotgun prompts "please enter your password again" 
- user hits cancel
- App X raises exception in maya script window
- User continues to work
- User launches App Y
- Shotgun prompts "please enter your password again" 
- etc

The idea is that pressing cancel doesn't alter the state of your user - toolkit will keep trying to authenticate you again and again. What happens when you run `uncache_session_token`?